packages/yarnpkg-pnp/sources/esm-loader/loader.ts

```diff
@@ -0,0 +1,135 @@
+import {ResolverFactory, CachedInputFileSystem} from 'enhanced-resolve';
+import fs                                       from 'fs';
+import {builtinModules}                         from 'module';
+import path                                     from 'path';
+import {fileURLToPath, pathToFileURL, URL}      from 'url';
+
+function isValidURL(str: string) {
+  try {
+    new URL(str);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+const builtins = new Set([...builtinModules]);
+
+const cachedFS = new CachedInputFileSystem(fs);
+
+function isDirectory(filePath: string) {
+  return new Promise<boolean>((resolve, reject) => {
+    cachedFS.lstat(filePath, (err, stat) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(stat.isDirectory());
+      }
+    });
+  });
+}
+
+function readFile(filePath: string) {
+  return new Promise<string>((resolve, reject) => {
+    cachedFS.readFile(filePath, (err, result) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(Buffer.isBuffer(result) ? result.toString(`utf8`) : result);
+      }
+    });
+  });
+}
+
+function readJson(filePath: string) {
+  return new Promise<any>((resolve, reject) => {
+    cachedFS.readJson(filePath, (err, result) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(result);
+      }
+    });
+  });
+}
+
+const commonResolver = ResolverFactory.createResolver({
+  fileSystem: cachedFS,
+  conditionNames: [`node`, `import`],
+  extensions: [`.mjs`, `.js`, `.json`],
+});
+
+export async function resolve(specifier: string, context: any, defaultResolver: any) {
+  if (builtins.has(specifier) || isValidURL(specifier)) return defaultResolver(specifier, context, defaultResolver);
+
+  const {parentURL, conditions = []} = context;
+
+  const resolver =
+    conditions.join(`.`) === `node.import`
+      ? commonResolver
+      : ResolverFactory.createResolver({
+        fileSystem: cachedFS,
+        conditionNames: conditions,
+        extensions: [`.mjs`, `.js`, `.json`],
```

Comment:
Should this also include `.cjs`?

Reply:
I think this should not include .mjs not .cjs since node.js usually doesn't append either of them by default. You need to explicit about them by adding them in require()

Reply:
Fair point, will remove